apis/dns/v1alpha1/dns_types.go

```diff
@@ -0,0 +1,397 @@
+
+	xpv1 "github.com/crossplane/crossplane-runtime/apis/common/v1"
+)
+
+// ZoneParameters define the desired state of an Azure DNS Zone.
+type ZoneParameters struct {
+	// ResourceGroupName specifies the name of the resource group that should
+	// contain this DNS Zone.
+	// +immutable
+	ResourceGroupName string `json:"resourceGroupName,omitempty"`
+
+	// ResourceGroupNameRef - A reference to a ResourceGroup object to retrieve
+	// its name
+	// +immutable
+	ResourceGroupNameRef *xpv1.Reference `json:"resourceGroupNameRef,omitempty"`
+
+	// ResourceGroupNameSelector - A selector for a ResourceGroup object to
+	// retrieve its name
+	// +immutable
+	ResourceGroupNameSelector *xpv1.Selector `json:"resourceGroupNameSelector,omitempty"`
+
+	// Location is the Azure location that the DNS Zone will be created in
+	// +required
+	// +immutable
+	Location string `json:"location"`
+
+	// ZoneType - Type of DNS zone to create.
+	// Allowed values: Private, Public
+	// Default: Public
+	// +kubebuilder:validation:Enum=Public;Private
+	// +kubebuilder:default=Public
+	// +optional
+	// +immutable
+	ZoneType string `json:"zoneType,omitempty"`
+
+	// RegistrationVirtualNetworks - A list of references to virtual networks that register hostnames in this
+	// DNS zone. This is an only when ZoneType is Private.
+	// +optional
+	RegistrationVirtualNetworks *[]SubResource `json:"registrationVirtualNetworks,omitempty"`
+
+	// ResolutionVirtualNetworks - A list of references to virtual networks that resolve records in this DNS zone.
+	// This is an only when ZoneType is Private.
+	// +optional
+	ResolutionVirtualNetworks *[]SubResource `json:"resolutionVirtualNetworks,omitempty"`
+
+	// Tags - Resource tags.
+	// +optional
+	Tags map[string]*string `json:"tags,omitempty"`
+}
+
+// ZoneObservation define the actual state of an Azure DNS Zone.
+type ZoneObservation struct {
+	// ID - Resource ID
+	ID string `json:"id,omitempty"`
+
+	// Etag - The etag of the zone.
+	Etag string `json:"etag,omitempty"`
+
+	// Name - The name of the zone.
+	Name string `json:"name,omitempty"`
+
+	// MaxNumberOfRecordSets - The maximum number of record sets that can be created in this DNS zone.
+	// This is a read-only property and any attempt to set this value will be ignored.
+	MaxNumberOfRecordSets int `json:"maxNumberOfRecordSets,omitempty"`
+
+	// NumberOfRecordSets - The current number of record sets in this DNS zone.
+	// This is a read-only property and any attempt to set this value will be ignored.
+	NumberOfRecordSets int `json:"numberOfRecordSets,omitempty"`
+
+	// NameServers - The name servers for this DNS zone. This is a read-only property and any attempt
+	// to set this value will be ignored.
+	NameServers []string `json:"nameServers,omitempty"`
+
+	// Type - Resource type.
+	Type string `json:"type,omitempty"`
+}
+
+// SubResource a reference to another resource
+type SubResource struct {
+	// ID - Resource id.
+	ID *string `json:"id,omitempty"`
```

Comment:
Why do we a `*string` here but a `string` type for `ZoneObservation.ID`?

Reply:
It seems that in SDK side, the `ID` field is not required for `SubResource`. So it is declared as a pointer. In repository, `ZoneObservation.ID` fields are string (are not pointer). So I applied this convention.